frontend-web/webclient/app/Files/LowLevelFileTable.tsx

```diff
@@ -398,6 +418,77 @@ export const LowLevelFileTable: React.FunctionComponent<LowLevelFileTableProps>
         setFileBeingRenamed(null);
     }, [Client.projectId]);
 
+    React.useEffect(() => {
+        if (wallet.data.items.length > 0) {
+            const s = new Set<string>();
+            wallet.data.items.forEach(it => s.add(it.category.provider));
+
+            fetchMachineSupport(
+                compute.jobs.retrieveProductsTemporary({
+                    providers: joinToString(Array.from(s), ",")
+            }));
+        }
+
+        if (selectedQuickLaunchApp !== null) {
+            fetchQuickLaunchApp(
+                compute.apps.findByNameAndVersion({
+                    appName: selectedQuickLaunchApp.metadata.name,
+                    appVersion: selectedQuickLaunchApp.metadata.version
+            }));
+        }
+    }, [wallet, selectedQuickLaunchApp]);
+
+
+    React.useEffect(() => {
+        if (quickLaunchApp.data !== null && props.path && machineSupport.data !== null) {
+            setAvailableProducts(
+                ([] as accounting.ProductNS.Compute[]).concat.apply(
+                    [],
+                    Object.values(machineSupport.data.productsByProvider).map(it => {
+                        return it.products
+                            .filter(it => {
+                                const tool = quickLaunchApp.data!!.invocation.tool.tool!;
+                                const backend = tool.description.backend;
+                                switch (backend) {
+                                    case "DOCKER":
+                                        return it.support.docker.enabled;
+                                    case "SINGULARITY":
+                                        return false;
+                                    case "VIRTUAL_MACHINE":
+                                        return it.support.virtualMachine.enabled &&
+                                            (tool.description.supportedProviders ?? [])
+                                                .some(p => p === it.product.category.provider);
+                                }
+                            })
+                            .filter(product =>
+                                wallet.data.items.some(wallet => productCategoryEquals(product.product.category, wallet.category))
+                            )
+                            .map(it => it.product);
+                    })
```

Comment:
I don't understand the usage of `apply` here. This could be simplified.

Reply:
We are essentially concatenating a list (providers) of lists (providers' products) into a list of products. I've removed the `concat.apply`. It works almost the same way now, but I would say that this is much more readable. If you have a better way to do this, please let me know.